src/Simulator/FluidSimulator.cpp

```diff
@@ -128,100 +128,95 @@ bool cFluidSimulator::IsHigherMeta(NIBBLETYPE a_Meta1, NIBBLETYPE a_Meta2)
 
 
 
-// TODO Not working very well yet :s
-Direction cFluidSimulator::GetFlowingDirection(int a_X, int a_Y, int a_Z, bool a_Over)
+Vector3f cFluidSimulator::GetFlowingDirection(int a_X, int a_Y, int a_Z)
 {
 	if (!cChunkDef::IsValidHeight(a_Y))
 	{
-		return NONE;
+		return {};
 	}
+
 	BLOCKTYPE BlockID = m_World.GetBlock(a_X, a_Y, a_Z);
-	if (!IsAllowedBlock(BlockID))  // No Fluid -> No Flowing direction :D
-	{
-		return NONE;
-	}
 
-	/*
-	Disabled because of causing problems and being useless atm
-	char BlockBelow = m_World.GetBlock(a_X, a_Y - 1, a_Z);  // If there is nothing or fluid below it -> dominating flow is down :D
-	if ((BlockBelow == E_BLOCK_AIR) || IsAllowedBlock(BlockBelow))
+	if (!IsAllowedBlock(BlockID))  // No Fluid -> No Flowing direction :D
 	{
-		return Y_MINUS;
+		return {};
 	}
-	*/
 
-	NIBBLETYPE LowestPoint = m_World.GetBlockMeta(a_X, a_Y, a_Z);  // Current Block Meta so only lower points will be counted
-	int X = 0, Z = 0;  // Lowest Pos will be stored here
+	NIBBLETYPE CentralPoint = m_World.GetBlockMeta(a_X, a_Y, a_Z);
 
-	if (IsAllowedBlock(m_World.GetBlock(a_X, a_Y + 1, a_Z)) && a_Over)  // check for upper block to flow because this also affects the flowing direction
+	if (CentralPoint > 7)
 	{
-		return GetFlowingDirection(a_X, a_Y + 1, a_Z, false);
+		CentralPoint -= 8;
 	}
 
-	std::vector< Vector3i * > Points;
-
-	Points.reserve(4);  // Already allocate 4 places :D
+	int LevelPoint[4];
 
-	// add blocks around the checking pos
-	Points.push_back(new Vector3i(a_X - 1, a_Y, a_Z));
-	Points.push_back(new Vector3i(a_X + 1, a_Y, a_Z));
-	Points.push_back(new Vector3i(a_X, a_Y, a_Z + 1));
-	Points.push_back(new Vector3i(a_X, a_Y, a_Z - 1));
+	// blocks around the checking pos
+	Vector3i Points[]
+	{
+		{ a_X + 1, a_Y, a_Z },
+		{ a_X, a_Y, a_Z + 1 },
+		{ a_X - 1, a_Y, a_Z },
+		{ a_X, a_Y, a_Z - 1 }
+	};
 
-	for (auto itr = Points.cbegin(), end = Points.cend(); itr != end; ++itr)
+	for (unsigned long i = 0; i < ARRAYCOUNT(LevelPoint); i++)
 	{
-		Vector3i * Pos = (*itr);
-		auto PosBlockID = m_World.GetBlock(Pos->x, Pos->y, Pos->z);
-		if (IsAllowedBlock(PosBlockID))
+		BlockID = m_World.GetBlock(Points[i].x, Points[i].y, Points[i].z);
+
+		if (IsAllowedBlock(BlockID))
 		{
-			NIBBLETYPE Meta = m_World.GetBlockMeta(Pos->x, Pos->y, Pos->z);
+			LevelPoint[i] = m_World.GetBlockMeta(Points[i].x, Points[i].y, Points[i].z);
 
-			if (Meta > LowestPoint)
+			if (LevelPoint[i] > 7)
+			{
+				LevelPoint[i] -= 8;
+			}
+			else if (LevelPoint[i] == 0)
 			{
-				LowestPoint = Meta;
-				X = Pos->x;
-				Z = Pos->z;
+				BlockID = m_World.GetBlock(Points[i].x, Points[i].y - 1, Points[i].z);
+
+				if (BlockID == m_FluidBlock)
+				{
+					LevelPoint[i] = CentralPoint + 1;
+				}
+				else if (BlockID == m_StationaryFluidBlock)
+				{
+					LevelPoint[i] = CentralPoint;
```

Comment:
This doesn't seem right. Don't stationary fluid blocks behaved exactly the same as the non-stationary versions?

Reply:
This eliminates the erratic flow in stationary water, try to run the game without this code and throw the object into a standing pond in which the bottom has a stepped depression.

Reply:
I'm not sure we're on the same page here.  A stationary water block is any water that isn't going to be updated by the fluid sim.  Stationary water can be flowing in any direction.

Also, could you expand on what you mean by "stepped depression"?

Reply:
But it's work good.

Reply:
Well, the vast majority of water is stationary so maybe the non-stationary fluid case is wrong.

Reply:
The following loop seems to work perfectly for me:
```cpp
for (size_t i = 0; i < ARRAYCOUNT(Points); i++)
{
	BlockID = m_World.GetBlock(Points[i].x, Points[i].y, Points[i].z);

	if (IsAllowedBlock(BlockID))
	{
		LevelPoint[i] = 0x07 & m_World.GetBlockMeta(Points[i].x, Points[i].y, Points[i].z);
	}
	else
	{
		LevelPoint[i] = CentralPoint;
	}
}
```

Reply:
Ok, I will try this.

Reply:
This is the wrong way.

Reply:
Sorry, it works. But we still have to test.

Reply:
Not so good. Water simulator Vanilla. 
![2018-01-21_11 18 56](https://user-images.githubusercontent.com/18181530/35190473-7c3d7c9e-fe9d-11e7-8b8d-eb450c67937c.png)
They must move.

Reply:
All the same, it works correctly.

Reply:
I can't seem to reproduce that, for me the diamonds are flowing as expected.

Reply:
It was my mistake, they also behave on the original server.

Reply:
Oh, OK.